File: Top-Interview-Questions/Strings/Easy-Collection/28-implement-strStr().py
class Solution:
    # Sol 1. Using Find built-in function
    def strStr(self, haystack: str, needle: str) -> int:
        
        return haystack.find(needle)
        # find returns -1 if no such string
        # index returns error if searching is not present
    # TC: O(n)
    # SC: O(1)
    # Runtime: 32ms, 58.97% faster
    # Memory: 14.1 MB, 12.31% less

    # Sol 2. Sliding Windows
    def strStr(self, haystack: str, needle: str) -> int:
        if len(needle) == 0:
            return 0 # corner case
        for i in range(len(haystack)): # traverse each index, len(heystack) - len(needle) + 1 will be faster
                                       # no need to look up the element after the haystack 
            if haystack[i:i+len(needle)] == needle:
                return i
        return -1
    # TC: O(n), where n is the length of haystack, worst case is that
    # we have to go thru all the element in haystack and needle length is one, and 
    # the index is located in the last poition of haystack
    # SC: O(n), no additional memories
    # Runtime: 36 ms, faster than 41.06%
    # Space: 13.9 MB, less than 12.31%

File: Top-Interview-Questions/Strings-Easy-Collection/28-implement-strStr().cpp
// Time Complexity: O((n-m+1) * m) (haystack length - needle length + 1) * (needle length)
// Space Complexity: O(m)
// Runtime: 12 ms, faster than 12.68%
// Memory Usage: 7.3 MB, less than 100.00%
class Solution {
public:
    bool checkSame(string s1, string s2){
        
        for(int i = 0; i < s1.length(); i++){
            if(s1[i] != s2[i])
                return false;
        }
            
        return true;
    }
    int strStr(string haystack, string needle) {
        
        if(!needle.length())
            return 0;
        
        if(needle.length() > haystack.length())
            return -1;
        
        string checkString;
        int start_index;
        
        for(int i = 0; i < haystack.length()-needle.length()+1; i++){
            
            start_index = i;
            checkString = "";
            
            for(int j = start_index; j < start_index + needle.length(); j++){
                checkString += haystack[j];
            }
            
            if(checkSame(checkString, needle))
                return start_index;
        }
        
        return -1;
    }
};
